Add hashing to A/B test variation assignment for true randomness.

aaronyan opened the pull request:
This PR adds hashing to the `assignmentMethod: 'userId'` for Calypso A/B tests. As detailed in #15169, the use of modulus on userids does not produce a truly random assignment across multiple tests. Using a hashing algorithm like MD5 or sha1 allows us to still use userId to determine the A/B test variation assigned, however, do so in a truly random fashion.

nylen commented:
A few thoughts:

### Bundle sizes, code complexity

From http://blog.richardweiss.org/2016/12/25/hash-splits.html:

> Hash them with a hashing algorithm (I use MD5, because it is also present in many DB systems)

I think this reasoning doesn't apply here, because using MD5 is "free" on the server, but not in the browser, because we have to ship a whole bunch of extra code to make it work.

The computation time is _probably_ negligible, but it would be worth looking at what this PR does to bundle sizes.  (And even if `crypto` is already included in the relevant bundles, that is a dependency that we should seek to remove, rather than relying on it more.)

A simpler alternative would be something like Java's `String.hashCode` where you loop over each character of the string, multiply the previous hash code value by a prime number, and add the code of each character to the hash value:

- https://docs.oracle.com/javase/6/docs/api/java/lang/String.html#hashCode()
- https://stackoverflow.com/a/299748

```js
const MAX_VALUE = 0xffffffff;
let hashCode = 0;
for ( let i = 0; i < str.length; i++ ) {
    hashCode = ( SOME_PRIME * hashCode + str.charCodeAt( i ) ) & MAX_VALUE;
}
```

Also, converting the user ID to a number then hashing it as a string will not end up using the full space of the hash function.  In practice this may not be a problem, but an alternative would be to process the user ID in chunks larger than decimal digits.  For example, to hash one byte at a time:

```js
let userIdRemainder = userId;
while ( userIdRemainder > 0 ) {
    hashCode = ( SOME_PRIME * hashCode + ( userIdRemainder & 0xff ) ) & MAX_VALUE;
    userIdRemainder >>= 8;
}
```

### Unit test coverage

There are undoubtedly a lot of edge cases to make sure that we handle correctly here (especially if we end up implementing our own hashing algorithm like the examples above, we need to be sure that our "fake integer math" behaves reasonably near the `MAX_VALUE` limit -- it [should be ok](https://stackoverflow.com/a/307200) but this needs to be tested first).

Any new code should be covered by unit tests, and ideally this critical module should be very well-covered: any code changes should cause test failures until the tests are updated as well.

It's probably also a good idea to test any new algorithm over a wide range of inputs and look for any patterns.

### Variant assignment algorithms

This is only semi-related to this PR, but I am not sure if there is a better place to discuss it.

I am not sure which A/B test algorithm we use more commonly, but when using the original random variant assigment, there are known problems with the same user being assigned to multiple variations (p10gg3-8o1-p2).

It seems like we should either start using this `userId` method for all A/B tests, or find a way to ensure that a given user is only assigned one variation of a given test.  One way to do this would be to do the variant assignment on the server using an API endpoint, then persist the A/B test variants in WP user meta.

nylen commented:
Shouldn't this PR introduce a new variant allocation algorithm like `userIdHashed`?  Otherwise, it seems like all existing tests that use the `userId` algorithm would be invalidated because their allocations would change.

jonathansadowski commented:
+1 to @nylen's comments about unit test coverage. Since the assignments should now be deterministic for a given userId / experimentId pair, it'd be great to see some tests in place to prevent this from being broken in the future.

Re: bundle-sizes... I agree this is an issue. I believe [`sha1` is already bundled](https://github.com/Automattic/wp-calypso/blob/5f15f3365213253724ca0d8798c6cb8160f00bc9/client/lib/wp/sync-handler/utils.js#L8), perhaps that'd be an option, but doesn't speak to the desire to remove that dependancy. Otherwise this gets to the discussion of whether or not we should be assigning variations on the client-side at all. I'm in favor of further discussion of loading existing variations from the backend (#20097). We could still randomly assign on the frontend so long as if an existing variation was in place it was reliably loaded prior to attempting random assignment.

> Shouldn't this PR introduce a new variant allocation algorithm like `userIdHashed`? Otherwise, it seems like all existing tests that use the `userId` algorithm would be invalidated because their allocations would change.

A quick search (`git log -SassignmentMethod`) shows that this assignment method has only ever been used once. Although, I agree that I'm in favor of a more explicit name, I don't think we need to worry about past tests, as they're all closed.

dmsnell commented:
http://iscalypsofastyet.com/branch?branch=update/analytics/ab-test-variation-by-userid

```
Delta:
chunk                                              stat_size            parsed_size            gzip_size
async-load-extensions-woocommerce-app-store-stats      +46 B   (+0.0%)        +73 B   (+0.0%)      +61 B   (+0.1%)
build                                              +576390 B   (+9.9%)    +324786 B  (+11.6%)   +98500 B  (+17.6%)
checkout                                           -141858 B  (-11.2%)     -58660 B   (-8.7%)   -15320 B  (-11.3%)
domains                                            -141858 B  (-11.2%)     -58680 B   (-8.3%)   -15491 B  (-12.2%)
jetpack-connect                                    -141858 B  (-21.0%)     -58586 B  (-16.3%)   -15107 B  (-19.9%)
manifest                                                +0 B                   +0 B                 -3 B   (-0.1%)
plans                                              -141858 B  (-23.9%)     -58581 B  (-18.5%)   -15154 B  (-23.1%)
purchases                                          -141858 B  (-13.8%)     -58609 B  (-10.8%)   -15131 B  (-13.9%)
settings                                           -141858 B  (-19.5%)     -58621 B  (-16.1%)   -15125 B  (-20.0%)
signup                                             -141858 B  (-15.0%)     -58588 B  (-11.4%)   -15410 B  (-13.7%)
vendor                                                 -24 B   (-0.0%)        -85 B   (-0.0%)      -49 B   (-0.0%)
```

Thanks to @jsnajdr we don't have to guess what the impact is on the bundles anymore! As we can see here, it would be EXTREMELY deleterious to incorporate this PR as-is because it would add 100KB to the gzip'd size of our initial bundle, increasing it's size by 17%!!!

jsnajdr commented on client/lib/abtest/index.js:
```diff
@@ -71,6 +71,8 @@ const parseDateStamp = datestamp => {
 const languageSlugs = map( config( 'languages' ), 'langSlug' );
 const langSlugIsValid = slug => languageSlugs.indexOf( slug ) !== -1;
 
+const crypto = require( 'crypto' );
```
Please don't use the `crypto` library -- it's a Node polyfill that's big and evil. As @jonathansadowski points out, there is a better option: `hash.js/lib/hash/sha/1` that's much much smaller. You'll need to use SHA1 instead of MD5, but that shouldn't matter at all in this case.

Using a simple prime-and-modulus hash algorithm that @nylen suggested is also a nice option. This feature doesn't need a hash that has any cryptographic properties.

aaronyan commented:
Thank you all so much for the comments and reviews @nylen, @jonathansadowski, @dmsnell, and @jsnajdr! I've replaced the `crypto` library with `sha1` and added a unit test! :)

aaronyan commented:
> I'm in favor of further discussion of loading existing variations from the backend (#20097).

aaronyan commented:
@jonathansadowski: for the broader discussion of A/B test variation assignment, I think we should try to always assign test variations deterministically (like using userid and hash functions). We then will never need to check the backend for previously assigned variations.

Our current hurdle is with anonymous users. Any thoughts on how to provide anonymous users the same test variation assignment?

dmsnell commented:
New bundle size diff is much better - thanks @aaronyan !

```
Delta:
chunk                                              stat_size           parsed_size           gzip_size
async-load-extensions-woocommerce-app-store-stats      +46 B  (+0.0%)        +73 B  (+0.0%)      +61 B  (+0.1%)
build                                                 +821 B  (+0.0%)       -401 B  (-0.0%)     -147 B  (-0.0%)
manifest                                                +0 B                  +0 B                -3 B  (-0.1%)
vendor                                                 -24 B  (-0.0%)        -85 B  (-0.0%)      -49 B  (-0.0%)
```

jsnmoon commented on client/lib/abtest/test/index.js:
```diff
@@ -143,6 +156,13 @@ describe( 'abtest', () => {
+				getUserStub.mockReturnValue( {
+					ID: '12321',
+				} );
+				expect( abtest( 'mockedTestAssignedWithUserID' ) ).to.equal( 'var3' );
```
To be honest, I have no idea what this test is actually doing...

kwight commented:
@aaronyan Are you still working on this? Looks like there are some code suggestions, but feel free to close if you don't think you'll be coming back to it here 👍

aaronyan commented:
Thank you for the bump @kwight! This PR was trying to fix #20097, but #22149 has already done so. Closing this PR for now, and I will start a new PR to remove the A/B test assignment by userid.
